vim-patch:8.2.{3510,3512,3514,3515,3534}: nanosecond timestamp change detection

zeertzjq opened the pull request:
#### vim-patch:8.2.3510: changes are only detected with one second accuracy

Problem:    Changes are only detected with one second accuracy.
Solution:   Use the nanosecond time if possible.  (Leah Neukirchen,
            closes vim/vim#8875)
https://github.com/vim/vim/commit/0a7984af5601323fae7b3398f05a48087db7b767

In Nvim Test_checktime_fast() is also flaky. Add a delay to avoid that.


#### vim-patch:8.2.3512: timestamp test fails on some systems

Problem:    Timestamp test fails on some systems.
Solution:   Sleep for a short while.
https://github.com/vim/vim/commit/accf4ed352c07ffe59022377c42d36e12dd6d461


#### vim-patch:8.2.3514: autoread test with nano second time sometimes fails

Problem:    Autoread test with nano second time sometimes fails.
Solution:   Mark the test as being flaky.
https://github.com/vim/vim/commit/eaa006dae3d5730e3b6dead27905444998b2cf8e


#### vim-patch:8.2.3515: nano time test fails on Mac and FreeBSD

Problem:    Nano time test fails on Mac and FreeBSD.
Solution:   Also check nano time when not on Linux. (Ozaki Kiichi,
            closes vim/vim#9000)
https://github.com/vim/vim/commit/def69dffb3d09a69629b071c89b7893a1783ba53


#### vim-patch:8.2.3534: autoread test is a bit flaky

Problem:    Autoread test is a bit flaky.
Solution:   Wait a brief moment before overwriting the file.
https://github.com/vim/vim/commit/944eeb44fb6e9d6d28474a1348d27c07873892f9

zeertzjq commented:
When I tested this locally, in Vim `Test_checktime_fast()` is not flaky, but in Nvim `Test_checktime_fast()` is flaky, and I have to add `sleep 10m` to make it less flaky. I'm not sure if this is normal.

zeertzjq commented:
Looks like `Test_checktime_fast()` is flaky in Nvim even if I set `'fsync'`.

seandewar commented on src/nvim/memline.c:
```diff
@@ -704,11 +704,13 @@ static void set_b0_fname(ZERO_BL *b0p, buf_T *buf)
       long_to_char(0L, b0p->b0_mtime);
       long_to_char(0L, b0p->b0_ino);
       buf->b_mtime = 0;
-      buf->b_mtime_read = 0;
```
Keep

seandewar reviewed (commented):
I'm also not sure what the cause of the flakiness is in Test_checktime_fast, but I think this is fine

seandewar reviewed (commented):
Apart from this review comment, I don't see any other issues.

The flakiness is weird though. Interestingly, I can't reproduce it locally with `fsync` set, even after running the test hundreds of times. It fails very consistently with it disabled (except in Vim), and it failed the CI that one time you tried it with it enabled. 🤷 As noted, Nvim's implementation is different after all.

zeertzjq commented on src/nvim/fileio.c:
```diff
@@ -3908,19 +3911,16 @@ static int check_mtime(buf_T *buf, FileInfo *file_info)
-/// @param t1 first time
-/// @param t2 second time
-static bool time_differs(long t1, long t2) FUNC_ATTR_CONST
+static bool time_differs(const uv_stat_t *st, long mtime, long mtime_ns) FUNC_ATTR_CONST
```
Indeed there is only one function that accepts `uv_stat_t`, but many that accpet `FileInfo`. I'll rename the argument to `file_info` then.

seandewar replied:
Yeah, generally `uv_stat_t` is not exposed outside of `os/file.c`.
Not sure why the `stat` field is accessed directly though in so many places, I recall that we had to use the `os_fileinfo_*` functions instead.

zeertzjq committed:
vim-patch:8.2.3510: changes are only detected with one second accuracy

Problem:    Changes are only detected with one second accuracy.
Solution:   Use the nanosecond time if possible.  (Leah Neukirchen,
            closes vim/vim#8873, closes vim/vim#8875)
https://github.com/vim/vim/commit/0a7984af5601323fae7b3398f05a48087db7b767

In Nvim Test_checktime_fast() is also flaky. Add a delay to avoid that.

zeertzjq committed:
vim-patch:8.2.3515: nano time test fails on Mac and FreeBSD

Problem:    Nano time test fails on Mac and FreeBSD.
Solution:   Also check nano time when not on Linux. (Ozaki Kiichi,
            closes vim/vim#9000)
https://github.com/vim/vim/commit/def69dffb3d09a69629b071c89b7893a1783ba53

zeertzjq committed:
vim-patch:8.2.3514: autoread test with nano second time sometimes fails

Problem:    Autoread test with nano second time sometimes fails.
Solution:   Mark the test as being flaky.
https://github.com/vim/vim/commit/eaa006dae3d5730e3b6dead27905444998b2cf8e

zeertzjq committed:
vim-patch:8.2.3534: autoread test is a bit flaky

Problem:    Autoread test is a bit flaky.
Solution:   Wait a brief moment before overwriting the file.
https://github.com/vim/vim/commit/944eeb44fb6e9d6d28474a1348d27c07873892f9

zeertzjq committed:
vim-patch:8.2.3512: timestamp test fails on some systems

Problem:    Timestamp test fails on some systems.
Solution:   Sleep for a short while.
https://github.com/vim/vim/commit/accf4ed352c07ffe59022377c42d36e12dd6d461
